gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala

```diff
@@ -674,62 +679,52 @@ case class TransformPostOverrides(isAdaptiveContext: Boolean) extends Rule[Spark
   }
 }
 
-object InMemoryTableScanHelper {
-  private def isGlutenTableCacheInternal(i: InMemoryTableScanExec): Boolean = {
-    // `ColumnarCachedBatchSerializer` is at velox module, so use class name here
-    i.relation.cacheBuilder.serializer.getClass.getSimpleName == "ColumnarCachedBatchSerializer" &&
-    i.supportsColumnar
-  }
-
-  def isGlutenTableCache(plan: SparkPlan): Boolean = {
-    plan match {
-      case i: InMemoryTableScanExec =>
-        isGlutenTableCacheInternal(i)
-      case q: QueryStageExec =>
-        // Compatible with Spark3.5 `TableCacheQueryStage`
-        isGlutenTableCache(q.plan)
-      case _ => false
-    }
-  }
-}
-
 // This rule will try to add RowToColumnarExecBase and ColumnarToRowExec
-// to support vanilla columnar scan.
+// to support vanilla columnar operators.
 case class VanillaColumnarPlanOverrides(session: SparkSession) extends Rule[SparkPlan] {
   @transient private val planChangeLogger = new PlanChangeLogger[SparkPlan]()
 
   private def replaceWithVanillaColumnarToRow(plan: SparkPlan): SparkPlan = plan match {
-    case c2r: ColumnarToRowExecBase if isVanillaColumnarReader(c2r.child) =>
+    case c2r: ColumnarToRowExecBase if PlanUtil.isVanillaColumnarOp(c2r.child) =>
       ColumnarToRowExec(c2r.child)
-    case c2r: ColumnarToRowExec if isVanillaColumnarReader(c2r.child) =>
+    case c2r: ColumnarToRowExec if PlanUtil.isVanillaColumnarOp(c2r.child) =>
       c2r
```

Comment:
Do we need to handle c2r's children recursively? Probably we can use `SparkPlan`'s `transformUp` API to avoid handwritten recursions.

Reply:
c2r cases can be removed, as they are used to prevent c2r is converted in original below case match. Now we don't need them.
Here we need to look two level plans: parent is gluten op and child is vanilla columnar op. I think transformUp may not be suitable.

Reply:
Sounds good to me. Thanks for the explanation.